packages/@aws-cdk/aws-codestarnotifications/lib/rule.ts

```diff
@@ -0,0 +1,262 @@
+
+  /**
+   * The events of notification rule
+   */
+  readonly events: string[] = [];
+
+  constructor(scope: Construct, id: string, props: RuleProps) {
+    super(scope, id);
+
+    this.source = this.bindSource(props.source);
+
+    if (props.events) {
+      this.addEvents(props.events);
+    }
+
+    props.targets?.forEach((target) => {
+      this.addTarget(target);
+    });
+
+    this.ruleName = props.ruleName || this.generateName();
+
+    this.ruleArn = new CfnNotificationRule(this, 'Resource', {
+      name: this.ruleName,
+      status: props?.status,
+      detailType: props.detailType || DetailType.FULL,
+      targets: this.targets,
+      eventTypeIds: this.events,
+      resource: this.source.sourceAddress,
+    }).ref;
+  }
+
+  /**
+   * Adds target to notification rule
+   * @param target The SNS topic or AWS Chatbot Slack target
+   */
+  public addTarget(target?: IRuleTarget): void {
+    if (!target) {
+      return;
+    }
+
+    this.targets.push(target.bind(this));
+  }
+
+  /**
+   * Adds events to notification rule
+   *
+   * @see https://docs.aws.amazon.com/dtconsole/latest/userguide/concepts.html#events-ref-pipeline
+   * @see https://docs.aws.amazon.com/dtconsole/latest/userguide/concepts.html#events-ref-buildproject
+   * @param events The list of event types for AWS Codebuild and AWS CodePipeline
+   */
+  public addEvents(events: string[]): void {
+    const re = {
+      [SourceType.CODE_BUILD]: /^codebuild\-/,
+      [SourceType.CODE_PIPELINE]: /^codepipeline\-/,
+    };
+
+    events.forEach((event) => {
+      if (!re[this.source.sourceType].test(event)) {
+        throw new Error(`${event} event id is not valid in ${this.source.sourceType}`);
+      }
+
+      if (this.events.includes(event)) {
+        return;
+      }
+
+      this.events.push(event);
+    });
+  }
+
+  private bindSource(source: IRuleSource): RuleSourceConfig {
+    const { projectArn, pipelineArn } = source;
+
+    // should throw error if multiple sources are specified
+    if ([projectArn, pipelineArn].filter(arn => arn !== undefined).length > 1) {
+      throw new Error('only one source can be specified');
+    }
+
+    if (projectArn) {
+      return {
+        sourceType: SourceType.CODE_BUILD,
+        sourceAddress: projectArn,
+      };
+    }
+
+    if (pipelineArn) {
+      return {
+        sourceType: SourceType.CODE_PIPELINE,
+        sourceAddress: pipelineArn,
+      };
+    }
+
+    throw new Error('"source" property must have "projectArn" or "pipelineArn"');
+  }
+
+  /**
+   * The name length generated by id must short than 64 characters
+   *
+   * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-codestarnotifications-notificationrule.html#cfn-codestarnotifications-notificationrule-name
+   * @returns string
+   */
+  private generateName(): string {
+    const name = Names.uniqueId(this);
+    if (name.length > 64) {
+      return name.substring(0, 63);
```

Comment:
You should take the last 64 characters for the name, not the first 64 ones.

Reply:
Resolved.